File: src/services/FDC3/channelClient.ts
declare global {
	interface Window {
		FSBL: typeof FSBL
	}
}

const win = window as Window;
export default class C implements Channel {
	id: string;
	type: string;
	displayMetadata?: DisplayMetadata;
	#FSBL: any;
	constructor(params: any) {
		this.id = params.id;
		this.type = params.type;
		this.displayMetadata = params.displayMetadata;
		this.#FSBL = win.FSBL || params.FSBL
	}

	broadcast(context: object): void {
		this.#FSBL.Clients.RouterClient.query(`FDC3.broadcast.${(context as any).type}`, {
			source: this.#FSBL.Clients.WindowClient.getWindowIdentifier().windowName, //used to prevent message loops
			channel: this.id,
			context
		}, () => { });
		this.#FSBL.Clients.RouterClient.query("FDC3.Channel.broadcast", {
			source: this.#FSBL.Clients.WindowClient.getWindowIdentifier().windowName, //used to prevent message loops
			channel: this.id,
			context
		}, () => { });
	}

	async getCurrentContext(contextType?: string): Promise<object> {
		const { err, response } = await this.#FSBL.Clients.RouterClient.query("FDC3.Channel.getCurrentContext", {
			channel: this.id,
			contextType
		}, () => { });
		if (err) {
			throw (err);
		} else {
			return response.data;
		}
	}

	addContextListener(handler: ContextHandler): Listener;
	addContextListener(contextType: string, handler: ContextHandler): Listener;
	addContextListener(contextTypeOrHandler: string | ContextHandler, handler?: ContextHandler): Listener {
		let theHandler: ContextHandler = null;
		let theListenerName: string = null;
		const currentWindowName = this.#FSBL.Clients.WindowClient.getWindowIdentifier().windowName

		//disambiguate arguments
		if (typeof contextTypeOrHandler === "string") {
			theHandler = handler;
			theListenerName = `FDC3.broadcast.${contextTypeOrHandler}`
		} else {
			theHandler = contextTypeOrHandler;
			theListenerName = `FDC3.broadcast`;
		}

		//only send the context data on if it did not get broadcast from this window
		const messageLoopPrevention = (_arg1: string | Error | null, { data }) =>
			data.source !== currentWindowName && theHandler(data.context)


		if (this.id == "global") {

			this.#FSBL.Clients.RouterClient.addListener(theListenerName, messageLoopPrevention);

			return {
				unsubscribe: () => {
					this.#FSBL.Clients.RouterClient.removeListener(theListenerName, messageLoopPrevention);
				}
			}

		} else {

			this.#FSBL.Clients.LinkerClient.linkToChannel(this.id, this.#FSBL.Clients.WindowClient.getWindowIdentifier());
			this.#FSBL.Clients.LinkerClient.subscribe(theListenerName, messageLoopPrevention);

			return {
				unsubscribe: () => {
					this.#FSBL.Clients.LinkerClient.unsubscribe(theListenerName, messageLoopPrevention);
				}
			}

		}
	}

}

File: dot-net-fdc3/FinsembleFDC3/ChannelClient.cs
namespace ChartIQ.Finsemble.FinsembleFDC3
{
	using Newtonsoft.Json.Linq;
	using Quobject.EngineIoClientDotNet.ComponentEmitter;
	using System;
	using System.Runtime.Remoting.Channels;

	public class ChannelClient : IChannel
	{
		internal const String ROUTER_FDC3_CHANNEL_BROADCAST = "FDC3.Channel.broadcast";
		internal const String ROUTER_FDC3_CHANNEL_GETCURRENTCONTEXT = "FDC3.Channel.getCurrentContext";
		internal const String FDC3_BROADCAST = "FDC3.broadcast";

		private Finsemble bridge;

		public String id { get; }

		public String type { get; }

		public DisplayMetadata displayMetadata { get; }

		DisplayMetadata IChannel.displayMetadata => throw new NotImplementedException();

		public ChannelClient(String id, String type, Finsemble bridge, DisplayMetadata displayMetadata)
		{
			this.id = id;
			this.type = type;
			this.bridge = bridge;
			this.bridge = bridge;
			this.displayMetadata = displayMetadata;
		}

		/// <summary>
		/// The broadcast.
		/// </summary>
		/// <param name="context">The context<see cref="JObject"/>.</param>
		public void broadcast(JObject context)
		{
			String windowName = this.bridge.WindowClient.windowIdentifier.GetValue("windowName").ToString();
			JObject param = new JObject() { 
				{ "source", windowName }, 
				{ "channel", this.id }, 
				{ context } 
			};
			this.bridge.RouterClient.Query(ROUTER_FDC3_CHANNEL_BROADCAST, param, (s, args) => { });
		}

		/// <summary>
		/// The broadcast.
		/// </summary>
		/// <param name="context">The context<see cref="JObject"/>.</param>
		public void getCurrentContext(String contextType, EventHandler<FinsembleEventArgs> responseHandler)
		{
			JObject param = new JObject { 
				{ "channel", this.id},
				{ contextType }
			};
			this.bridge.RouterClient.Query(ROUTER_FDC3_CHANNEL_GETCURRENTCONTEXT, param, responseHandler);
		}

		/// <summary>
		/// The broadcast.
		/// </summary>
		/// <param name="context">The context<see cref="JObject"/>.</param>
		public IListener addContextListener(String contextType, IContextHandler handler) {
			String windowName = this.bridge.WindowClient.windowIdentifier.GetValue("windowName").ToString();
			String theListenerName = "";

			if(contextType != null)
			{
				theListenerName = FDC3_BROADCAST + "." + contextType;
			} else
			{
				theListenerName = FDC3_BROADCAST;
			}

			if (this.id.Equals("global"))
			{
				EventHandler<FinsembleEventArgs> routerHandler = (sender, args) => {
					var response = (JObject)args.response;
					String source = ((JValue)response?["data"]?["source"])?.ToString();
					if (!source.Equals(windowName))
					{
						handler.handle((JObject)response?["data"]?["context"]);
					}
				};
				this.bridge.RouterClient.AddListener(theListenerName, routerHandler);
				return new RouterListener(this.bridge, theListenerName, routerHandler);
			}
			else
			{
				EventHandler<FinsembleEventArgs> linkerHandler = (sender, args) => {
					var response = (JObject)args.response;
					String source = ((JValue)response?["source"])?.ToString();
					if (!source.Equals(windowName))
					{
						handler.handle((JObject)response?["data"]?["context"]);
					}
				};
				this.bridge.LinkerClient.LinkToChannel(this.id, this.bridge.WindowClient.windowIdentifier, (sender, args) => { });

				this.bridge.LinkerClient.Subscribe(theListenerName, linkerHandler);
				return new LinkerListener(this.bridge, theListenerName, linkerHandler);
			}
		}
	}
}
